Andrej/update client

Andrej123 opened the pull request:
Implemented update client feature.
Also fixed a bug that prevented the deletion of clients, as Client.id is optional.

Andrej123 committed:
Implemented #12 - It should be possible to update the client.

Andrej123 committed:
Fixed bug that prevented the deletion of clients.

goranche commented:
I'm not sure I like the "show an alert with a textfield" approach... wouldn't it be better to have inline edit in the table view cell itself?

Andrej123 commented:
Ok, agree, I'll change that. I'll make use of UITextField inside customised cells. I'll experiment a little, we'll see how it'll work :)

goranche commented:
@Andrej123 so we can expect a talk on the topic for the next meetup? 😇 
I think it would be an interesting thing 👍

Andrej123 committed:
Updating client name inline implemented using textField delegate as separate class.

Andrej123 committed:
Added reload(:) to ClientsDataSource.swift. Change so that we're not always creating a new object in refresh(:), but instead just reload the clients array.

goranche commented on chargeOclock/Source/DataSources/ClientsDataSource.swift:
```diff
@@ -27,10 +32,9 @@ class ClientsDataSource: NSObject, UITableViewDataSource {
-		if var clientCell = cell as? ClientSettable {
-			clientCell.client = clients[indexPath.row]
-		}
+		let cell = tableView.dequeueReusableCell(withIdentifier: reuseIdentifier, for: indexPath) as! ClientTableViewCell
```
remove the forced cast, use the pattern that was used before, since it's much more robust.

goranche commented on chargeOclock/Source/DataSources/ClientsDataSource.swift:
```diff
@@ -11,13 +11,18 @@ import UIKit
 		self.reuseIdentifier = reuseIdentifier
 	}
 
+	func reload(_ clients: [Client]) {
```
why couldn't you just set the clients array?
and add a didSet handler if anything needs to be done if the array changes.

goranche commented on chargeOclock/Source/Views/ClientTableViewCell.swift:
```diff
@@ -10,16 +10,32 @@ import UIKit
+		guard let client = client else {
+			return
+		}
+		updateClient?(client)
```
why do we need this? couldn't you just call save on the client object?

goranche commented on chargeOclock/devTemp/ClientTextFieldDelegate.swift:
```diff
@@ -0,0 +1,26 @@
+
+import UIKit
+
+class ClientTextFieldDelegate: NSObject, UITextFieldDelegate {
```
not sure I'd create a separate class just for this... and even if, it would need a better name, since the current name doesn't give a clue as to what it does (except that it's a text field delegate)

Andrej123 committed:
Removed ClientTextFieldDelegate as a separate class and extended ClientTableViewCell to act as UITextFieldDelegate.
In ClientsDatasSource removed force casting to ClientTableViewCell and using ClientSettable protocol to set client propertie.
Removed unused updateClient(:) in devClientsViewController.swift.

goranche commented on chargeOclock/Source/DataSources/ClientsDataSource.swift:
```diff
@@ -32,9 +27,10 @@ class ClientsDataSource: NSObject, UITableViewDataSource {
-		let cell = tableView.dequeueReusableCell(withIdentifier: reuseIdentifier, for: indexPath) as! ClientTableViewCell
-		cell.client = clients[indexPath.row]
-		cell.updateClient = updateClient
+		let cell = tableView.dequeueReusableCell(withIdentifier: reuseIdentifier, for: indexPath)
```
why the separate variable? you could do a:
`if var cell = tableView.deque... {`

Andrej123 replied:
The issue is that ClientSettable is just a protocol defining that there must be a *client* property. If I try this:
```
if var cell = tableView.dequeueReusableCell(withIdentifier: reuseIdentifier, for: indexPath) as? ClientSettable {
    cell.client = clients[indexPath.row]
    return cell
}
```
I get an error stating that I can't return a ClientSettable, but instead UITableViewCell type is expected. So I've changed it to this:
```
let cell = tableView.dequeueReusableCell(withIdentifier: reuseIdentifier, for: indexPath)
    if var cell = cell as? ClientSettable {
        cell.client = clients[indexPath.row]
    }
return cell
```
But I'm not 100% sure that in this way is ok, as I think the changed variant only works because cell is a reference type property (class, not structure which is value type). I'm not sure it's ok, as if cell was a structure, this code wouldn't work.

goranche replied:
🤔 ok

goranche commented on chargeOclock/Source/Views/ClientTableViewCell.swift:
```diff
@@ -13,29 +13,51 @@ class ClientTableViewCell: UITableViewCell, ClientSettable {
+		client.name = text
+		client.save { error in
+			guard error == nil else {
+				let alert = UIAlertController(title: "Save error", message: "Client saving failed, error \(error?.localizedDescription)", preferredStyle: .alert)
```
maybe change `, error ` into `with error:\n`, will probably look better 😉

goranche commented on chargeOclock/Source/Views/ClientTableViewCell.swift:
```diff
@@ -13,29 +13,51 @@ class ClientTableViewCell: UITableViewCell, ClientSettable {
+				let alert = UIAlertController(title: "Save error", message: "Client saving failed, error \(error?.localizedDescription)", preferredStyle: .alert)
+				alert.addAction(UIAlertAction(title: "Ok", style: .default, handler: nil ))
+
+				var parentViewController: UIViewController? {
```
while this will work, it doesn't really make sense to have this as a calculated property, so just make it into a variable and perform the same operations on it

edit: ah, sorry... after looking through the code in more detail, it is needed as is... ignore this comment 😇

goranche commented on chargeOclock/Source/Views/ClientTableViewCell.swift:
```diff
@@ -13,29 +13,51 @@ class ClientTableViewCell: UITableViewCell, ClientSettable {
+				alert.addAction(UIAlertAction(title: "Ok", style: .default, handler: nil ))
+
+				var parentViewController: UIViewController? {
+					var parentResponder: UIResponder? = self
```
since we "know" we're not a view controller, you might as well start with `self.next`, saves one iteration

goranche commented on chargeOclock/Source/Views/ClientTableViewCell.swift:
```diff
@@ -13,29 +13,51 @@ class ClientTableViewCell: UITableViewCell, ClientSettable {
+					var parentResponder: UIResponder? = self
+					while parentResponder != nil {
+						parentResponder = parentResponder!.next
+						if parentResponder is UIViewController {
```
to avoid the forced cast, change this to:
```
if let parentResponder = parentResponder as? UIViewController {
    return parentResponder
}
```

goranche reviewed (changes requested):
not bad, I think it looks much cleaner... a few minor details remain, but overall 👍

Andrej123 committed:
Cleanup related to CR.

goranche commented on chargeOclock/Source/Views/ClientTableViewCell.swift:
```diff
@@ -41,20 +41,19 @@ extension ClientTableViewCell: UITextFieldDelegate {
 		client.save { error in
 			guard error == nil else {
-				let alert = UIAlertController(title: "Save error", message: "Client saving failed, error \(error?.localizedDescription)", preferredStyle: .alert)
+				let alert = UIAlertController(title: "Save error", message: "Client saving failed, with error:\n\(error)", preferredStyle: .alert)
```
why did you leave out `localizedDescription`?

Andrej123 replied:
I left it out as I found that `localizedDescription` doesn't work as I thought. I found out that using just the error the displayed text corresponds to the name of the error case (eg. *internalError*, *serverError(401)*, ...). In contrast, using the `localizedDescription` displayed some unfamiliar, useless description.

But I did some search now and I see (I guess 🤔) the "right" way should be to use `localizedDescription` and also implement the `localizedDescription` computed property in `oClockError`.

Andrej123 committed:
Implemented `localizedDescription` for oClockerError and added again `localizedDescription` in ClientTableViewCell.
